README
                               nss-myhostname 0.3

   Copyright 2005-2011 Lennart Poettering <mzzlubfganzr (at) 0pointer
   (dot) de>
     * [1]License
     * [2]News
     * [3]Overview
     * [4]Current Status
     * [5]Documentation
     * [6]Requirements
     * [7]Installation
     * [8]Acknowledgements
     * [9]Download

License

   This program is free software; you can redistribute it and/or modify it
   under the terms of the GNU Lesser General Public License as published
   by the Free Software Foundation, either version 2.1 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser
   General Public License for more details.

News

   Mon May 9 2011:

   [10]Version 0.3 released. Changes include: Always return locally
   configured IP addresses first, use 127.0.0.2/::1 only as fallback if no
   addresses are configured.

   Mon Oct 27 2008:

   [11]Version 0.2 released. Changes include: Update for IPv6 and newest
   glibc NSS interfaces

   Sat Nov 19 2005:

   [12]Version 0.1 released. Initial release.

Overview

   nss-myhostname is a plugin for the GNU Name Service Switch (NSS)
   functionality of the GNU C Library (glibc) providing host name
   resolution for the locally configured system hostname as returned by
   gethostname(2). Various software relies on an always resolvable local
   host name. When using dynamic hostnames this is usually achieved by
   patching /etc/hosts at the same time as changing the host name. This
   however is not ideal since it requires a writable /etc file system and
   is fragile because the file might be edited by the administrator at the
   same time. nss-myhostname simply returns all locally configure public
   IP addresses, or -- if none are configured -- the IPv4 address
   127.0.0.2 (wich is on the local loopback) and the IPv6 address ::1
   (which is the local host) for whatever system hostname is configured
   locally. Patching /etc/hosts is thus no longer necessary.

Current Status

   It works!

Documentation

   After compiling and installing nss-myhostname you'll find a new NSS
   modules in /lib:
     * libnss_myhostname.so.2

   To activate the NSS modules you have to edit /etc/nsswitch.conf and add
   myhostname to the line starting with "hosts:". On Debian this looks
   like this:
# /etc/nsswitch.conf

passwd:         compat
group:          compat
shadow:         compat

hosts:          files dns myhostname
networks:       files

protocols:      db files
services:       db files
ethers:         db files
rpc:            db files

netgroup:       nis

   That's it. You should now always be able to resolve your local system
   hostname. For a quick check use glibc's getent tool:
$ getent ahosts `hostname`
::1             STREAM omega
::1             DGRAM
::1             RAW
127.0.0.2       STREAM
127.0.0.2       DGRAM
127.0.0.2       RAW

   In this case the local host name is omega.

   It is recommended to put myhostname last in the nsswitch.conf line to
   make sure that this mapping is only used as fallback, and any DNS or
   /etc/hosts based mapping takes precedence.

Requirements

   nss-myhostname uses NSS interfaces that are specific to modern glibc's.

   nss-myhostname was developed and tested on Fedora 15 from May 2011, it
   should work on most other Linux distributions since it uses GNU
   autoconf and GNU libtool for source code configuration and shared
   library management.

Installation

   As this package is made with the GNU autotools you should run
   ./configure inside the distribution directory for configuring the
   source tree. After that you should run make for compilation and make
   install (as root) for installation of nss-myhostname.

Acknowledgements

   None so far.

Download

   The newest release is always available from
   [13]http://0pointer.de/lennart/projects/nss-myhostname/

   The current release is [14]0.3

   Get nss-myhostname's development sources from the [15]GIT
   [16]repository ([17]gitweb):
git clone git://git.0pointer.de/nss-myhostname
     __________________________________________________________________


    Lennart Poettering <mzzlubfganzr (at) 0pointer (dot) de>, May 2011

References

   1. README#license
   2. README#news
   3. README#overview
   4. README#status
   5. README#documentation
   6. README#requirements
   7. README#installation
   8. README#acks
   9. README#download
  10. http://0pointer.de/lennart/projects/nss-myhostname/nss-myhostname-0.3.tar.gz
  11. http://0pointer.de/lennart/projects/nss-myhostname/nss-myhostname-0.2.tar.gz
  12. http://0pointer.de/lennart/projects/nss-myhostname/nss-myhostname-0.1.tar.gz
  13. http://0pointer.de/lennart/projects/nss-myhostname/
  14. http://0pointer.de/lennart/projects/nss-myhostname/nss-myhostname-0.3.tar.gz
  15. http://git.or.cz/
  16. git://git.0pointer.de/nss-myhostname
  17. http://git.0pointer.de/?p=nss-myhostname.git


debian/changelog
libnss-myhostname (0.3-9) unstable; urgency=medium

  * s/172/127/g in patch names and comments.

 -- Joachim Breitner <nomeata@debian.org>  Wed, 23 Jul 2014 13:05:31 +0200

libnss-myhostname (0.3-8) unstable; urgency=medium

  * Remove toc from README.html (Closes: #755788) 

 -- Joachim Breitner <nomeata@debian.org>  Wed, 23 Jul 2014 12:32:29 +0200

libnss-myhostname (0.3-7) unstable; urgency=medium

  * Add autopkgtest to test hostname resolution without /etc/hosts, and also
    test with nscd installed.
  * Fix nscd compatibility (Closes: #755053), thank to Sam Morris for digging
    up the patch, taking from the systemd repository.

 -- Joachim Breitner <nomeata@debian.org>  Tue, 22 Jul 2014 18:01:52 +0200

libnss-myhostname (0.3-6) unstable; urgency=low

  * Fix duplicated ldconfig calls in maintainer scripts, thanks to Michael
    Prokop for noticing. (Closes: #712802)

 -- Joachim Breitner <nomeata@debian.org>  Thu, 20 Jun 2013 11:06:06 +0200

libnss-myhostname (0.3-5) unstable; urgency=low

  * Work around #705900: Ignore link-local addresses

 -- Joachim Breitner <nomeata@debian.org>  Tue, 21 May 2013 15:43:56 +0200

libnss-myhostname (0.3-4) unstable; urgency=low

  * Fix Package's description to reflect changed behaviour. (Closes: #656218)

 -- Joachim Breitner <nomeata@debian.org>  Tue, 17 Jan 2012 22:08:50 +0100

libnss-myhostname (0.3-3) unstable; urgency=low

  * Implement Multi-Arch support 
  * 0004-Ensure-a-LANG-independent-generated-README.patch: 
    Make sure the generated html file is independent of the current LANG

 -- Joachim Breitner <nomeata@debian.org>  Tue, 28 Jun 2011 12:53:59 +0200

libnss-myhostname (0.3-2) unstable; urgency=low

  * Apply patch by Robert Millan, adding legacy support for BSD and hurd
    kernels (Closes: #630206) 
  * Use dh-autoreconf to take care of autoreconfing 

 -- Joachim Breitner <nomeata@debian.org>  Fri, 17 Jun 2011 09:53:40 +0200

libnss-myhostname (0.3-1) unstable; urgency=low

  * Fix typo in package description (Closes: #624526)
  * New upstream release (Closes: #630009)
  * Bump standard version, no change
  * Install myshostname directly after "files" in /etc/nsswitch.conf.
    This avoids an annoying delay when dns fails and one wants to use sudo to
    fix it. Is also closer to having the hostname in /etc/hosts, which is the
    behaviour myhostname tries to mimic. Existing installations are not
    modified.

 -- Joachim Breitner <nomeata@debian.org>  Fri, 10 Jun 2011 23:04:56 +0200

libnss-myhostname (0.2-5) unstable; urgency=low

  * Bump standards version, no change
  * Do not run dh_shlibps for .so files in /lib32 or /lib64, avoids dependency
    on (e.g.) libc6-amd64 (Closes: #587918)

 -- Joachim Breitner <nomeata@debian.org>  Wed, 07 Jul 2010 17:31:34 +0200

libnss-myhostname (0.2-4) unstable; urgency=low

  * Try to pass a more correct --build parameter to ./configure when building
    multilib variants.

 -- Joachim Breitner <nomeata@debian.org>  Mon, 21 Jun 2010 14:58:27 +0200

libnss-myhostname (0.2-3) unstable; urgency=low

  * Typo in description (Closes: #584640)
  * Build-Depend on lynx for README creation
  * Actually use -m64 option when building 64bit libraries

 -- Joachim Breitner <nomeata@debian.org>  Sun, 20 Jun 2010 23:16:34 +0200

libnss-myhostname (0.2-2) unstable; urgency=low

  * Fix VCS URL in controls file 
  * Use 127.0.1.1 for consistency with Debian’s default /etc/hosts (Closes:
    #583624)
  * Amend configure.ac to cater for non-defined MAX_HOST_NAME

 -- Joachim Breitner <nomeata@debian.org>  Fri, 04 Jun 2010 12:59:21 +0200

libnss-myhostname (0.2-1) unstable; urgency=low

  * Initial release (Closes: #574800)

 -- Joachim Breitner <nomeata@debian.org>  Thu, 06 May 2010 15:34:50 +0200


legacy.c
/*-*- Mode: C; c-basic-offset: 8; indent-tabs-mode: nil -*-*/

/***
  This file is part of nss-myhostname.

  Copyright 2008-2011 Lennart Poettering
  Copyright 2011 Robert millan

  nss-myhostname is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public License
  as published by the Free Software Foundation; either version 2.1 of
  the License, or (at your option) any later version.

  nss-myhostname is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with nss-myhostname; If not, see
  <http://www.gnu.org/licenses/>.
***/

#include <sys/types.h>
#include <errno.h>
#include <ifaddrs.h>
#include <stdlib.h>
#include <string.h>
#include <netinet/in.h>

#include "ifconf.h"

int ifconf_acquire_addresses(struct address **_list, unsigned *_n_list) {

        struct address *list = NULL;
        unsigned n_list = 0;

        struct ifaddrs *ifa = NULL;
        int r = 1;
        struct ifaddrs *i;
        int ifindex = 0;

        if (getifaddrs(&ifa) == -1) { 
                r = -errno;
                goto finish;
        }

        for (i = ifa; i != NULL; i = i->ifa_next) {
                int af;
                const void *cp;
                struct sockaddr_in6 *in6 = (struct sockaddr_in6 *) i->ifa_addr;
                struct sockaddr_in *in = (struct sockaddr_in *) i->ifa_addr;

                if (! i->ifa_addr)
                        continue;

                af = i->ifa_addr->sa_family;

                if (af != AF_INET && af != AF_INET6)
                        continue;

                list = realloc(list, (n_list+1) * sizeof(struct address));
                if (!list) {
                        r = -ENOMEM;
                        goto finish;
                }

                if (af == AF_INET6)
                        cp = &in6->sin6_addr;
                else
                        cp = &in->sin_addr;

                list[n_list].family = af;
                list[n_list].scope = 0;
                memcpy(list[n_list].address, cp, af == AF_INET ? 4 : 16);
                list[n_list].ifindex = ifindex++;
                n_list++;
        }

finish:
        if (ifa)
                freeifaddrs(ifa);

        if (r < 0)
                free(list);
        else {
                qsort(list, n_list, sizeof(struct address), address_compare);

                *_list = list;
                *_n_list = n_list;
        }

        return r;
}


nss-myhostname.c
/*-*- Mode: C; c-basic-offset: 8; indent-tabs-mode: nil -*-*/

/***
  This file is part of nss-myhostname.

  Copyright 2008-2011 Lennart Poettering

  nss-myhostname is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public License
  as published by the Free Software Foundation; either version 2.1 of
  the License, or (at your option) any later version.

  nss-myhostname is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with nss-myhostname; If not, see
  <http://www.gnu.org/licenses/>.
***/

#include <limits.h>
#include <nss.h>
#include <sys/types.h>
#include <netdb.h>
#include <errno.h>
#include <string.h>
#include <assert.h>
#include <unistd.h>
#include <net/if.h>
#include <stdlib.h>
#include <arpa/inet.h>

#include "ifconf.h"

/* We use 127.0.0.2 as IPv4 address. This has the advantage over
 * 127.0.0.1 that it can be translated back to the local hostname. For
 * IPv6 we use ::1 which unfortunately will not translate back to the
 * hostname but instead something like "localhost6" or so. */

/* DEBIAN: Default /etc/hosts uses 172.0.1.1, so that is what we use as well */

#define LOCALADDRESS_IPV4 (htonl(0x7F000101))
#define LOCALADDRESS_IPV6 &in6addr_loopback
#define LOOPBACK_INTERFACE "lo"

#define ALIGN(a) (((a+sizeof(void*)-1)/sizeof(void*))*sizeof(void*))

enum nss_status _nss_myhostname_gethostbyname4_r(
                const char *name,
                struct gaih_addrtuple **pat,
                char *buffer, size_t buflen,
                int *errnop, int *h_errnop,
                int32_t *ttlp) _public_;

enum nss_status _nss_myhostname_gethostbyname3_r(
                const char *name,
                int af,
                struct hostent *host,
                char *buffer, size_t buflen,
                int *errnop, int *h_errnop,
                int32_t *ttlp,
                char **canonp) _public_;

enum nss_status _nss_myhostname_gethostbyname2_r(
                const char *name,
                int af,
                struct hostent *host,
                char *buffer, size_t buflen,
                int *errnop, int *h_errnop) _public_;

enum nss_status _nss_myhostname_gethostbyname_r(
                const char *name,
                struct hostent *host,
                char *buffer, size_t buflen,
                int *errnop, int *h_errnop) _public_;

enum nss_status _nss_myhostname_gethostbyaddr2_r(
                const void* addr, socklen_t len,
                int af,
                struct hostent *host,
                char *buffer, size_t buflen,
                int *errnop, int *h_errnop,
                int32_t *ttlp) _public_;

enum nss_status _nss_myhostname_gethostbyaddr_r(
                const void* addr, socklen_t len,
                int af,
                struct hostent *host,
                char *buffer, size_t buflen,
                int *errnop, int *h_errnop) _public_;

enum nss_status _nss_myhostname_gethostbyname4_r(
                const char *name,
                struct gaih_addrtuple **pat,
                char *buffer, size_t buflen,
                int *errnop, int *h_errnop,
                int32_t *ttlp) {

        unsigned lo_ifi;
        char hn[HOST_NAME_MAX+1];
        size_t l, idx, ms;
        char *r_name;
        struct gaih_addrtuple *r_tuple, *r_tuple_prev = NULL;
        struct address *addresses = NULL, *a;
        unsigned n_addresses = 0, n;

        memset(hn, 0, sizeof(hn));
        if (gethostname(hn, sizeof(hn)-1) < 0) {
                *errnop = errno;
                *h_errnop = NO_RECOVERY;
                return NSS_STATUS_UNAVAIL;
        }

        if (strcasecmp(name, hn) != 0) {
                *errnop = ENOENT;
                *h_errnop = HOST_NOT_FOUND;
                return NSS_STATUS_NOTFOUND;
        }

        /* If this fails, n_addresses is 0. Which is fine */
        ifconf_acquire_addresses(&addresses, &n_addresses);

        /* If this call fails we fill in 0 as scope. Which is fine */
        lo_ifi = if_nametoindex(LOOPBACK_INTERFACE);

        l = strlen(hn);
        ms = ALIGN(l+1)+ALIGN(sizeof(struct gaih_addrtuple))*(n_addresses > 0 ? n_addresses : 2);
        if (buflen < ms) {
                *errnop = ENOMEM;
                *h_errnop = NO_RECOVERY;
                free(addresses);
                return NSS_STATUS_TRYAGAIN;
        }

        /* First, fill in hostname */
        r_name = buffer;
        memcpy(r_name, hn, l+1);
        idx = ALIGN(l+1);

        if (n_addresses <= 0) {
                /* Second, fill in IPv6 tuple */
                r_tuple = (struct gaih_addrtuple*) (buffer + idx);
                r_tuple->next = r_tuple_prev;
                r_tuple->name = r_name;
                r_tuple->family = AF_INET6;
                memcpy(r_tuple->addr, LOCALADDRESS_IPV6, 16);
                r_tuple->scopeid = (uint32_t) lo_ifi;

                idx += ALIGN(sizeof(struct gaih_addrtuple));
                r_tuple_prev = r_tuple;

                /* Third, fill in IPv4 tuple */
                r_tuple = (struct gaih_addrtuple*) (buffer + idx);
                r_tuple->next = r_tuple_prev;
                r_tuple->name = r_name;
                r_tuple->family = AF_INET;
                *(uint32_t*) r_tuple->addr = LOCALADDRESS_IPV4;
                r_tuple->scopeid = (uint32_t) lo_ifi;

                idx += ALIGN(sizeof(struct gaih_addrtuple));
                r_tuple_prev = r_tuple;
        }

        /* Fourth, fill actual addresses in, but in backwards order */
        for (a = addresses + n_addresses - 1, n = 0; n < n_addresses; n++, a--) {
                r_tuple = (struct gaih_addrtuple*) (buffer + idx);
                r_tuple->next = r_tuple_prev;
                r_tuple->name = r_name;
                r_tuple->family = a->family;
                r_tuple->scopeid = a->ifindex;
                memcpy(r_tuple->addr, a->address, 16);

                idx += ALIGN(sizeof(struct gaih_addrtuple));
                r_tuple_prev = r_tuple;
        }

        /* Verify the size matches */
        assert(idx == ms);

        /* Nscd expects us to store the first record in **pat. */
        if (*pat)
                **pat = *r_tuple_prev;
        else
                *pat = r_tuple_prev;

        if (ttlp)
                *ttlp = 0;

        free(addresses);

        return NSS_STATUS_SUCCESS;
}

static enum nss_status fill_in_hostent(
                const char *hn,
                int af,
                struct hostent *result,
                char *buffer, size_t buflen,
                int *errnop, int *h_errnop,
                int32_t *ttlp,
                char **canonp) {

        size_t l, idx, ms;
        char *r_addr, *r_name, *r_aliases, *r_addr_list;
        size_t alen;
        struct address *addresses = NULL, *a;
        unsigned n_addresses = 0, n, c;

        alen = PROTO_ADDRESS_SIZE(af);

        ifconf_acquire_addresses(&addresses, &n_addresses);

        for (a = addresses, n = 0, c = 0; n < n_addresses; a++, n++)
                if (af == a->family)
                        c++;

        l = strlen(hn);
        ms = ALIGN(l+1)+
                sizeof(char*)+
                (c > 0 ? c : 1)*ALIGN(alen)+
                (c > 0 ? c+1 : 2)*sizeof(char*);

        if (buflen < ms) {
                *errnop = ENOMEM;
                *h_errnop = NO_RECOVERY;
                free(addresses);
                return NSS_STATUS_TRYAGAIN;
        }

        /* First, fill in hostname */
        r_name = buffer;
        memcpy(r_name, hn, l+1);
        idx = ALIGN(l+1);

        /* Second, create (empty) aliases array */
        r_aliases = buffer + idx;
        *(char**) r_aliases = NULL;
        idx += sizeof(char*);

        /* Third, add addresses */
        r_addr = buffer + idx;
        if (c > 0) {
                unsigned i = 0;

                for (a = addresses, n = 0; n < n_addresses; a++, n++) {
                        if (af != a->family)
                                continue;

                        memcpy(r_addr + i*ALIGN(alen), a->address, alen);
                        i++;
                }

                assert(i == c);
                idx += c*ALIGN(alen);
        } else {
                if (af == AF_INET)
                        *(uint32_t*) r_addr = LOCALADDRESS_IPV4;
                else
                        memcpy(r_addr, LOCALADDRESS_IPV6, 16);

                idx += ALIGN(alen);
        }

        /* Fourth, add address pointer array */
        r_addr_list = buffer + idx;
        if (c > 0) {
                unsigned i = 0;

                for (a = addresses, n = 0; n < n_addresses; a++, n++) {
                        if (af != a->family)
                                continue;

                        ((char**) r_addr_list)[i] = (r_addr + i*ALIGN(alen));
                        i++;
                }

                assert(i == c);
                ((char**) r_addr_list)[c] = NULL;
                idx += (c+1)*sizeof(char*);

        } else {
                ((char**) r_addr_list)[0] = r_addr;
                ((char**) r_addr_list)[1] = NULL;
                idx += 2*sizeof(char*);
        }

        /* Verify the size matches */
        assert(idx == ms);

        result->h_name = r_name;
        result->h_aliases = (char**) r_aliases;
        result->h_addrtype = af;
        result->h_length = alen;
        result->h_addr_list = (char**) r_addr_list;

        if (ttlp)
                *ttlp = 0;

        if (canonp)
                *canonp = r_name;

        free(addresses);

        return NSS_STATUS_SUCCESS;
}

enum nss_status _nss_myhostname_gethostbyname3_r(
                const char *name,
                int af,
                struct hostent *host,
                char *buffer, size_t buflen,
                int *errnop, int *h_errnop,
                int32_t *ttlp,
                char **canonp) {

        char hn[HOST_NAME_MAX+1];

        if (af == AF_UNSPEC)
                af = AF_INET;

        if (af != AF_INET && af != AF_INET6) {
                *errnop = EAFNOSUPPORT;
                *h_errnop = NO_DATA;
                return NSS_STATUS_UNAVAIL;
        }

        memset(hn, 0, sizeof(hn));
        if (gethostname(hn, sizeof(hn)-1) < 0) {
                *errnop = errno;
                *h_errnop = NO_RECOVERY;
                return NSS_STATUS_UNAVAIL;
        }

        if (strcasecmp(name, hn) != 0) {
                *errnop = ENOENT;
                *h_errnop = HOST_NOT_FOUND;
                return NSS_STATUS_NOTFOUND;
        }

        return fill_in_hostent(hn, af, host, buffer, buflen, errnop, h_errnop, ttlp, canonp);
}

enum nss_status _nss_myhostname_gethostbyname2_r(
                const char *name,
                int af,
                struct hostent *host,
                char *buffer, size_t buflen,
                int *errnop, int *h_errnop) {

        return _nss_myhostname_gethostbyname3_r(
                        name,
                        af,
                        host,
                        buffer, buflen,
                        errnop, h_errnop,
                        NULL,
                        NULL);
}

enum nss_status _nss_myhostname_gethostbyname_r(
                const char *name,
                struct hostent *host,
                char *buffer, size_t buflen,
                int *errnop, int *h_errnop) {

        return _nss_myhostname_gethostbyname3_r(
                        name,
                        AF_UNSPEC,
                        host,
                        buffer, buflen,
                        errnop, h_errnop,
                        NULL,
                        NULL);
}

enum nss_status _nss_myhostname_gethostbyaddr2_r(
                const void* addr, socklen_t len,
                int af,
                struct hostent *host,
                char *buffer, size_t buflen,
                int *errnop, int *h_errnop,
                int32_t *ttlp) {

        char hn[HOST_NAME_MAX+1];
        struct address *addresses = NULL, *a;
        unsigned n_addresses = 0, n;

        if (len != PROTO_ADDRESS_SIZE(af)) {
                *errnop = EINVAL;
                *h_errnop = NO_RECOVERY;
                return NSS_STATUS_UNAVAIL;
        }

        if (af == AF_INET) {

                if ((*(uint32_t*) addr) == LOCALADDRESS_IPV4)
                        goto found;

        } else if (af == AF_INET6) {

                if (memcmp(addr, LOCALADDRESS_IPV6, 16) == 0)
                        goto found;

        } else {
                *errnop = EAFNOSUPPORT;
                *h_errnop = NO_DATA;
                return NSS_STATUS_UNAVAIL;
        }

        ifconf_acquire_addresses(&addresses, &n_addresses);

        for (a = addresses, n = 0; n < n_addresses; n++, a++) {
                if (af != a->family)
                        continue;

                if (memcmp(addr, a->address, PROTO_ADDRESS_SIZE(af)) == 0)
                        goto found;
        }

        *errnop = ENOENT;
        *h_errnop = HOST_NOT_FOUND;

        free(addresses);
        return NSS_STATUS_NOTFOUND;

found:
        free(addresses);

        memset(hn, 0, sizeof(hn));
        if (gethostname(hn, sizeof(hn)-1) < 0) {
                *errnop = errno;
                *h_errnop = NO_RECOVERY;

                return NSS_STATUS_UNAVAIL;
        }

        return fill_in_hostent(hn, af, host, buffer, buflen, errnop, h_errnop, ttlp, NULL);

}

enum nss_status _nss_myhostname_gethostbyaddr_r(
                const void* addr, socklen_t len,
                int af,
                struct hostent *host,
                char *buffer, size_t buflen,
                int *errnop, int *h_errnop) {

        return _nss_myhostname_gethostbyaddr2_r(
                        addr, len,
                        af,
                        host,
                        buffer, buflen,
                        errnop, h_errnop,
                        NULL);
}


netlink.c
/*-*- Mode: C; c-basic-offset: 8; indent-tabs-mode: nil -*-*/

/***
  This file is part of nss-myhostname.

  Copyright 2008-2011 Lennart Poettering

  nss-myhostname is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public License
  as published by the Free Software Foundation; either version 2.1 of
  the License, or (at your option) any later version.

  nss-myhostname is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with nss-myhostname; If not, see
  <http://www.gnu.org/licenses/>.
***/

#include <sys/socket.h>
#include <sys/un.h>
#include <asm/types.h>
#include <inttypes.h>
#include <linux/netlink.h>
#include <linux/rtnetlink.h>
#include <string.h>
#include <assert.h>
#include <errno.h>
#include <limits.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <inttypes.h>
#include <stdlib.h>

#include "ifconf.h"

int ifconf_acquire_addresses(struct address **_list, unsigned *_n_list) {

        struct {
                struct nlmsghdr hdr;
                struct rtgenmsg gen;
        } req;
        struct rtgenmsg *gen;
        int fd, r, on = 1;
        uint32_t seq = 4711;
        struct address *list = NULL;
        unsigned n_list = 0;

        fd = socket(PF_NETLINK, SOCK_DGRAM, NETLINK_ROUTE);
        if (fd < 0)
                return -errno;

        if (setsockopt(fd, SOL_SOCKET, SO_PASSCRED, &on, sizeof(on)) < 0) {
                r = -errno;
                goto finish;
        }

        memset(&req, 0, sizeof(req));
        req.hdr.nlmsg_len = NLMSG_LENGTH(sizeof(struct rtgenmsg));
        req.hdr.nlmsg_type = RTM_GETADDR;
        req.hdr.nlmsg_flags = NLM_F_REQUEST|NLM_F_DUMP|NLM_F_ACK;
        req.hdr.nlmsg_seq = seq;
        req.hdr.nlmsg_pid = 0;

        gen = NLMSG_DATA(&req.hdr);
        gen->rtgen_family = AF_UNSPEC;

        if (send(fd, &req, req.hdr.nlmsg_len, 0) < 0) {
                r = -errno;
                goto finish;
        }

        for (;;) {
                ssize_t bytes;
                struct msghdr msg;
                struct cmsghdr *cmsg;
                struct ucred *ucred;
                struct iovec iov;
                struct nlmsghdr *p;
                uint8_t cred_buffer[CMSG_SPACE(sizeof(struct ucred))];
                struct {
                        struct nlmsghdr hdr;
                        struct ifaddrmsg ifaddrmsg;
                        uint8_t payload[16*1024];
                } resp;

                memset(&iov, 0, sizeof(iov));
                iov.iov_base = &resp;
                iov.iov_len = sizeof(resp);

                memset(&msg, 0, sizeof(msg));
                msg.msg_name = NULL;
                msg.msg_namelen = 0;
                msg.msg_iov = &iov;
                msg.msg_iovlen = 1;
                msg.msg_control = cred_buffer;
                msg.msg_controllen = sizeof(cred_buffer);
                msg.msg_flags = 0;

                bytes = recvmsg(fd, &msg, 0);
                if (bytes < 0) {
                        r = -errno;
                        goto finish;
                }

                cmsg = CMSG_FIRSTHDR(&msg);
                if (!cmsg || cmsg->cmsg_type != SCM_CREDENTIALS) {
                        r = -EIO;
                        goto finish;
                }

                ucred = (struct ucred*) CMSG_DATA(cmsg);
                if (ucred->uid != 0 || ucred->pid != 0)
                        continue;

                for (p = &resp.hdr; bytes > 0; p = NLMSG_NEXT(p, bytes)) {
                        struct ifaddrmsg *ifaddrmsg;
                        struct rtattr *a;
                        size_t l;
                        void *local = NULL, *address = NULL;

                        if (!NLMSG_OK(p, (size_t) bytes)) {
                                r = -EIO;
                                goto finish;
                        }

                        if (p->nlmsg_seq != seq)
                                continue;

                        if (p->nlmsg_type == NLMSG_DONE) {
                                r = 0;
                                goto finish;
                        }

                        if (p->nlmsg_type == NLMSG_ERROR) {
                                struct nlmsgerr *nlmsgerr;

                                nlmsgerr = NLMSG_DATA(p);
                                r = -nlmsgerr->error;
                                goto finish;
                        }

                        if (p->nlmsg_type != RTM_NEWADDR)
                                continue;

                        ifaddrmsg = NLMSG_DATA(p);

                        if (ifaddrmsg->ifa_family != AF_INET &&
                            ifaddrmsg->ifa_family != AF_INET6)
                                continue;

                        if (ifaddrmsg->ifa_scope == RT_SCOPE_HOST ||
                            ifaddrmsg->ifa_scope == RT_SCOPE_NOWHERE)
                                continue;

                        if (ifaddrmsg->ifa_flags & IFA_F_DEPRECATED)
                                continue;

                        l = NLMSG_PAYLOAD(p, sizeof(struct ifaddrmsg));
                        a = IFA_RTA(ifaddrmsg);

                        while (RTA_OK(a, l)) {

                                if (a->rta_type == IFA_ADDRESS)
                                        address = RTA_DATA(a);
                                else if (a->rta_type == IFA_LOCAL)
                                        local = RTA_DATA(a);

                                a = RTA_NEXT(a, l);
                        }

                        if (local)
                                address = local;

                        if (!address)
                                continue;

			// Avoid link-local address
			// http://bugs.debian.org/705900
			if (ifaddrmsg->ifa_family == AF_INET6 &&
			    IN6_IS_ADDR_LINKLOCAL((const struct in6_addr *)address))
			    continue;

                        list = realloc(list, (n_list+1) * sizeof(struct address));
                        if (!list) {
                                r = -ENOMEM;
                                goto finish;
                        }

                        list[n_list].family = ifaddrmsg->ifa_family;
                        list[n_list].scope = ifaddrmsg->ifa_scope;
                        memcpy(list[n_list].address, address, ifaddrmsg->ifa_family == AF_INET ? 4 : 16);
                        list[n_list].ifindex = ifaddrmsg->ifa_index;

                        n_list++;
                }
        }

finish:
        close(fd);

        if (r < 0)
                free(list);
        else {
                qsort(list, n_list, sizeof(struct address), address_compare);

                *_list = list;
                *_n_list = n_list;
        }

        return r;
}
